docs/plugins.rst

```diff
@@ -0,0 +1,101 @@
+
+
+Overall architecture flow
+*************************
+In order to create a plugin which will work with ape, you will need to 
+    * define a class that subclasses the abstract methods within the ``ape`` api.
+    * implement all the methods in order for it to work
+
+Types of plugins
+================
+* 1st - These are plugins that are bundled with ape core. They are built in, don't have a version, and can't be uninstallled.
+* 2nd - These plugins are maintained by ApeWorX. They are trusted. Users and developers can pin different versions but should be aware of api changes when doing so.
+* 3rd - These are plugins developed outside the ApeWorX organization. These should be installed at your own risk. These can also be pinned like in 2nd order plugins.
+... TBD could be referenced in the user documentation or possibly live in the user documentation
+
+
+Plugin Registration process flow
+================================
+Ape uses ``pluggy`` for plugin management. The ``@plugins.register`` decorator hooks into ape core. 
+The plugin process looks for all local installed site packages that start with ``ape_``.
+The plugin process will loop through these potential ape plugins and see which ones have created a plugin type registration.
+If the plugin type registration is found, then ``ape`` knows that this package is a plugin and attempts to process it according to registration interface. 
+Then we have a set of registered plugins that the registration process defines it needs. The ``@hookspec`` decorator describes how the plugin works. 
+CLI registration
+
+
+Compilation process flow
+========================
+The project manager object is a representation of your current project, which should contain all the files the user's project will use, including ``contracts/`` folder (where contract source code is stored).
+The ``contracts/`` folder is where the compiler looks for contracts to compile.
+File extensions found within the ``contracts/`` directory determine which compiler plugin ape uses.
+The pragma spec of the compilable files within the folder is checked and then used to decide if a new compiler needs to be 
+downloaded or if the version matches one of the currently installed compiler versions. 
+The contracts are then grouped by compiler type and version and fed into the corresponding compiler to compile them. 
+These are then output as a JSON file to the ``.build`` directory. They can then be deployed on the chain from the console or a script.
+
+
+Source type
+    Types from types module :doc:`source types <autoapi/ape/types/contract/index.html#ape.types.contract.Source>`
+
+Compiler manager
+    The compiler manager contains all the registered compilers. 
+    Compilers subclass the compiler API object, implementing the CompilerAPI methods.
+Implement :doc:`get_versions <autoapi/ape/api/compiler/index#ape.api.compiler.CompilerAPI.get_versions>` in compile.
+CompilerAPI plugins
+
+ContractType type
```

Comment:
link to autoapi

Reply:
the json is the manifest file

Reply:
Contract types: data and properties 
The manifest is a file that describes the package. It describes everything that is within the package. The package manifest contains links to source code on ipfs, a heirarchy of which compilers and which versions of those compilers are used to compile files, contract types which come from the compiled files that you might want to use in the package. 
From the contract types you can then initialize or deploy a contract using the contract types.

Reply:
data structure source is how to get the source file from the manifest

Reply:
Compiler Manager uses the compiler api and has a list of all the compiler api subclasses. The compiler manager has the set of all the registered compiler plugins. Those compiler plugins subclass the compiler api, and so that's how it can call out to the plugins in order to compile files which are detected inside the contracts folder.

Reply:
added this all